File: Shell_Sort/Shell_Sort.cpp
// C++ implementation of Shell Sort
#include <iostream>

using namespace std;

int Shell_Sort(int array[], int size)
{
    for(int gap = size / 2; gap > 0; gap /= 2)
    {
        for(int i = gap; i < size; i++)
        {
            int temp = array[i], j;

            for(j = i; j >= gap && array[j - gap] > temp; j -= gap)
                array[j] = array[j - gap];

            array[j] = temp;
        }
    }

    return 0;
}

void Print_Array(int array[], int size)
{
    for(int i = 0; i < size; i++)
        cout << array[i] << " ";

    cout << endl;
}

int main()
{
    int array[] = {12, 34, 54, 2, 3};
    int size = 5;

    Shell_Sort(array, size);
    Print_Array(array, size);

    return 0;
}


/* Output

2 3 12 34 54

*/

File: Shell_Sort/Shell_Sort.java
/* Java implementation of Shell Sort*/

public class Shell_Sort
{
    public void sort(int input[]) // Function implementing Insertion Sort
    {
          int n = input.length, temp, j;
          for (int gap = n / 2; gap > 0; gap /= 2)
          {
              for (int i = gap; i < n; i++) // Gapped insertion sort
              {
                  temp = input[i];

                  for (j = i; j >= gap && input[j - gap] > temp; j = j - gap)
                      input[j] = input[j - gap];

                  input[j] = temp;
              }
          }
    }

    public static void main(String[] args)
    {
        int input[] = {5, 6, 8, 4, 6, 5, 3, 3, 2, 7, 8, 45, 85, 96}; //Input array

        shellSort ob = new shellSort();

        ob.sort(input);

        System.out.print("Output : ");

        for (int i = 0; i < input.length; ++i) // Printing Sorted array
            System.out.print(input[i] + " ");
    }

}

/*

Output : 2 3 3 4 5 5 6 6 7 8 8 45 85 96

*/
